Add unboxed instances for all remaining base newtypes

Shimuuar opened the pull request:
At some point I needed unboxed vector of `Down Double` opnly to discover that it's not available so I added all Unbox instances for all newtypes in base

 - Down
 - Dual
 - Sum
 - Product
 - All & Any

RyanGlScott commented:
Thanks for the pull request, @Shimuuar.

It's worth noting that if your goal here is to literally cover every single newtype in `base`, this PR alone doesn't accomplish that, since (OtToMH) there are more newtypes in:

* `Data.Monoid` (`Alt`)
* `Data.Functor.Compose` (`Compose`)
* `Data.Semigroup` (several)

That being said, perhaps you don't want to bother with those newtypes, which would be fine by me. In that case, I'd simply request that you fix the compilation error on GHC 7.4, which is caused by `Down` not being exported by `Data.Ord` in that version of GHC. You could fix this by importing `Down` from `GHC.Exts` instead, since it's been exported from there since even before 7.4.

Shimuuar commented:
I added instances for types your mentioned and changed import for `Down`. Hopefully everything will work now but CI will tell

Shimuuar commented:
Also something pulled base-orpahns which provided Foldab;e/Travesable instances for tuples and broke tests. I removed orphans and added explicit dependency.

Shimuuar commented:
Hmm... CI fails intermittently when trying to install cabal & GHC.

cartazio commented:
... wait ... I’m not sure if vector can depend on base orphans/should it?

I guess that’s fine. But @ryangiscott is the one who has clarity on that

On Wed, Jan 3, 2018 at 11:47 AM Aleksey Khudyakov <notifications@github.com>
wrote:

> Hmm... CI fails intermittently when trying to install cabal & GHC.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/vector/pull/194#issuecomment-355061562>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwltjobC7hhGDFv10D1UgynJMfnTzks5tG68BgaJpZM4RPnsG>
> .
>

RyanGlScott commented:
To be clear, it's the test suite that has a `base-orphans` dependency, which I think is entirely reasonable.

I'm not sure why the Travis build is failing, on the other hand. I'll have to dig deeper to see what is going on there.

cartazio commented:
Ok that sounds good

On Wed, Jan 3, 2018 at 8:29 PM Ryan Scott <notifications@github.com> wrote:

> To be clear, it's the test suite that has a base-orphans dependency,
> which I think is entirely reasonable.
>
> I'm not sure why the Travis build is failing, on the other hand. I'll have
> to dig deeper to see what is going on there.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/vector/pull/194#issuecomment-355175433>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwnKfBhKxSFy6xCy737VQMI9FR_PEks5tHClrgaJpZM4RPnsG>
> .
>

RyanGlScott commented:
Sorry for taking so long to get back to this, @Shimuuar. Can you rebase this patch on top of `master`? I've applied some patches recently to make the Travis configuration a bit more reliable w.r.t. testing.

RyanGlScott commented on Data/Vector/Unboxed/Base.hs:
```diff
@@ -412,11 +421,129 @@ instance (Unbox a) => G.Vector Vector (Complex a) where
-instance (Unbox a) => M.MVector MVector (Identity a) where
+newtype instance MVector s (Down a) = MV_Down (MVector s a)
+newtype instance Vector (Down a) = V_Down (Vector a)
+instance (Unbox a) => Unbox (Down a)
```
As a general thought, is it possible to have the `newtypeMVector`/`newtypeVector` macros generate these `MVector`, `Vector`, and `Unbox` instances? That might save a good deal of boilerplate.

Shimuuar commented:
I rebased it and added another CPP macro to generate newtype instances. 

Although if you use CPP to generate code something went horribly wrong at some point. I wonder is it possible to better support for working data that is not stored uniformly (boxed) to haskell?

RyanGlScott commented:
Indeed, using CPP is a bit unfortunate, but it's likely the best tool we have at the moment (as we likely don't want to introduce, say, a Template Haskell dependency in `vector`).

There are likely other ways we could cut down on the boilerplate. For instance, [these](https://github.com/Shimuuar/vector/blob/a5197aeb4825226ff6cbc8f7ad102bcdfff13107/Data/Vector/Unboxed/Base.hs#L426-L450) two [chunks](https://github.com/Shimuuar/vector/blob/a5197aeb4825226ff6cbc8f7ad102bcdfff13107/Data/Vector/Unboxed/Base.hs#L558-L582) of code are exactly the same, so we could define a standalone CPP macro to consolidate the two. (If you want to do this, that would be great, but otherwise we could do this later.)

Other than that, I think this looks good to me. What say you, @cartazio?

cartazio commented:
I’ll eyeball it later today or tomorrow.  Hit me up on irc tomorrow
afternoon if I’ve not done so by then :)

On Fri, Jan 19, 2018 at 2:33 PM Ryan Scott <notifications@github.com> wrote:

> Indeed, using CPP is a bit unfortunate, but it's likely the best tool we
> have at the moment (as we likely don't want to introduce, say, a Template
> Haskell dependency in vector).
>
> There are likely other ways we could cut down on the boilerplate. For
> instance, these
> <https://github.com/Shimuuar/vector/blob/a5197aeb4825226ff6cbc8f7ad102bcdfff13107/Data/Vector/Unboxed/Base.hs#L426-L450>
> two chunks
> <https://github.com/Shimuuar/vector/blob/a5197aeb4825226ff6cbc8f7ad102bcdfff13107/Data/Vector/Unboxed/Base.hs#L558-L582>
> of code are exactly the same, so we could define a standalone CPP macro to
> consolidate the two. (If you want to do this, that would be great, but
> otherwise we could do this later.)
>
> Other than that, I think this looks good to me. What say you, @cartazio
> <https://github.com/cartazio>?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/vector/pull/194#issuecomment-359066808>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwmTERRsFV7t5N_q95YmTpxur9r67ks5tMO4LgaJpZM4RPnsG>
> .
>

cartazio commented:
It'd be nice if the CPP also was setup to play nice with the multi param arg newtypes, but perhaps thats wishful thinking on my part?

or perhaps the nice one would be on top of a  "cruder" one that handles Arg and Compose and friends?

Shimuuar commented:
I factored out macro for MVector instance for newtypes. But for all other instances there're slight variations and they're too awkward to encode in CPP. In my opinion CPP is just too awkward and error prone to try to encode even somewhat complicated logic in it. Boilerplate is probably preferable

cartazio commented:
I’ll see if I can help you with this.  Vector really doesn’t need more
boilerplate. :)

On Mon, Jan 22, 2018 at 4:12 AM Aleksey Khudyakov <notifications@github.com>
wrote:

> I factored out macro for MVector instance for newtypes. But for all other
> instances there're slight variations and they're too awkward to encode in
> CPP. In my opinion CPP is just too awkward and error prone to try to encode
> even somewhat complicated logic in it. Boilerplate is probably preferable
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/vector/pull/194#issuecomment-359364085>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwvpzPAwf31dYIeN2jweLumiLT-Rzks5tNFDrgaJpZM4RPnsG>
> .
>

Shimuuar commented:
I just remembered about this PR and looked into travis log. Line 

```
cabal new-build -w ${HC} ${TEST} ${BENCH} ${TARGETS}
```

was SIGKILL'ed for 7.10, 8.2 and 8.4. Presumable because it runned out of memory but travis provides no info

RyanGlScott commented:
@Shimuuar, try rebasing on top of `master`—I tweaked the Travis setup recently to avoid this issue.

Shimuuar commented:
@RyanGlScott thank it indeed fixed build

RyanGlScott commented:
Thanks, @Shimuuar. One suggestion to reduce the size of the Diff: you should be able to derive several of the instances that you currently write out by hand using a single macro. I believe this patch should do the trick, no?

```diff
diff --git a/Data/Vector/Unboxed/Base.hs b/Data/Vector/Unboxed/Base.hs
index 226d98b..12146cb 100644
--- a/Data/Vector/Unboxed/Base.hs
+++ b/Data/Vector/Unboxed/Base.hs
@@ -421,8 +421,8 @@ instance (Unbox a) => G.Vector Vector (Complex a) where
 -- -------
 -- Identity
 -- -------
-#define newtypeMVector(ty,tyC,con) \
-instance (Unbox a) => M.MVector MVector (ty a) where { \
+#define newtypeMVector(inst_ctxt,inst_head,tyC,con) \
+instance inst_ctxt => M.MVector MVector (inst_head) where { \
 ; {-# INLINE basicLength          #-}                                         \
 ; {-# INLINE basicUnsafeSlice     #-}                                         \
 ; {-# INLINE basicOverlaps        #-}                                         \
@@ -449,8 +449,8 @@ instance (Unbox a) => M.MVector MVector (ty a) where { \
 ; basicUnsafeMove (con v1) (con v2)  = M.basicUnsafeMove v1 v2                \
 ; basicUnsafeGrow (con v) n          = con `liftM` M.basicUnsafeGrow v n      \
 }
-#define newtypeVector(ty,tyC,con,mcon) \
-instance (Unbox a) => G.Vector Vector (ty a) where { \
+#define newtypeVector(inst_ctxt,inst_head,tyC,con,mcon) \
+instance inst_ctxt => G.Vector Vector (inst_head) where { \
 ; {-# INLINE basicUnsafeFreeze  #-}                                       \
 ; {-# INLINE basicUnsafeThaw    #-}                                       \
 ; {-# INLINE basicLength        #-}                                       \
@@ -465,23 +465,21 @@ instance (Unbox a) => G.Vector Vector (ty a) where { \
 ; basicUnsafeCopy (mcon mv) (con v) = G.basicUnsafeCopy mv v              \
 ; elemseq _ (tyC a)                 = G.elemseq (undefined :: Vector a) a \
 }
-#define deriveNewtypeInstances(ty,tyC,con,mcon) \
-newtype instance MVector s (ty a) = mcon (MVector s a) ;\
-newtype instance Vector    (ty a) = con  (Vector a)    ;\
-instance (Unbox a) => Unbox (ty a)                     ;\
-newtypeMVector(ty, tyC, mcon)                          ;\
-newtypeVector(ty, tyC, con, mcon)
-
-
+#define deriveNewtypeInstances(inst_ctxt,inst_head,rep,tyC,con,mcon) \
+newtype instance MVector s (inst_head) = mcon (MVector s (rep)) ;\
+newtype instance Vector    (inst_head) = con  (Vector (rep))    ;\
+instance inst_ctxt => Unbox (inst_head)                         ;\
+newtypeMVector(inst_ctxt, inst_head, tyC, mcon)                 ;\
+newtypeVector(inst_ctxt,  inst_head, tyC, con, mcon)
 
 #if MIN_VERSION_base(4,8,0)
-deriveNewtypeInstances(Identity, Identity, V_Identity, MV_Identity)
+deriveNewtypeInstances(Unbox a, Identity a, a, Identity, V_Identity, MV_Identity)
 #endif
 
-deriveNewtypeInstances(Down,    Down,    V_Down,    MV_Down)
-deriveNewtypeInstances(Dual,    Dual,    V_Dual,    MV_Dual)
-deriveNewtypeInstances(Sum,     Sum,     V_Sum,     MV_Sum)
-deriveNewtypeInstances(Product, Product, V_Product, MV_Product)
+deriveNewtypeInstances(Unbox a, Down a,    a, Down,    V_Down,    MV_Down)
+deriveNewtypeInstances(Unbox a, Dual a,    a, Dual,    V_Dual,    MV_Dual)
+deriveNewtypeInstances(Unbox a, Sum a,     a, Sum,     V_Sum,     MV_Sum)
+deriveNewtypeInstances(Unbox a, Product a, a, Product, V_Product, MV_Product)
 
 
 -- --------------
@@ -489,11 +487,11 @@ deriveNewtypeInstances(Product, Product, V_Product, MV_Product)
 -- --------------
 
 #if MIN_VERSION_base(4,9,0)
-deriveNewtypeInstances(Min,   Min,   V_Min,   MV_Min)
-deriveNewtypeInstances(Max,   Max,   V_Max,   MV_Max)
-deriveNewtypeInstances(First, First, V_First, MV_First)
-deriveNewtypeInstances(Last,  Last,  V_Last,  MV_Last)
-deriveNewtypeInstances(WrappedMonoid, WrapMonoid, V_WrappedMonoid, MV_WrappedMonoid)
+deriveNewtypeInstances(Unbox a, Min a,   a, Min,   V_Min,   MV_Min)
+deriveNewtypeInstances(Unbox a, Max a,   a, Max,   V_Max,   MV_Max)
+deriveNewtypeInstances(Unbox a, First a, a, First, V_First, MV_First)
+deriveNewtypeInstances(Unbox a, Last a,  a, Last,  V_Last,  MV_Last)
+deriveNewtypeInstances(Unbox a, WrappedMonoid a, a, WrapMonoid, V_WrappedMonoid, MV_WrappedMonoid)
 
 -- ------------------
 -- Data.Semigroup.Arg
@@ -549,168 +547,21 @@ instance (Unbox a, Unbox b) => G.Vector Vector (Arg a b) where
                                  $ G.elemseq (undefined :: Vector b) y z
 #endif
 
--- NOTE!
---
--- `elemseq' here is implemented as `seq' That's OK for Bool
--- underlying type
-#define newtypeMVectorMono(ty,con) \
-instance M.MVector MVector ty where { \
-; {-# INLINE basicLength          #-}                                        \
-; {-# INLINE basicUnsafeSlice     #-}                                        \
-; {-# INLINE basicOverlaps        #-}                                        \
-; {-# INLINE basicUnsafeNew       #-}                                        \
-; {-# INLINE basicInitialize      #-}                                        \
-; {-# INLINE basicUnsafeReplicate #-}                                        \
-; {-# INLINE basicUnsafeRead      #-}                                        \
-; {-# INLINE basicUnsafeWrite     #-}                                        \
-; {-# INLINE basicClear           #-}                                        \
-; {-# INLINE basicSet             #-}                                        \
-; {-# INLINE basicUnsafeCopy      #-}                                        \
-; {-# INLINE basicUnsafeGrow      #-}                                        \
-; basicLength (con v)               = M.basicLength v                        \
-; basicUnsafeSlice i n (con v)      = con $ M.basicUnsafeSlice i n v         \
-; basicOverlaps (con v1) (con v2)   = M.basicOverlaps v1 v2                  \
-; basicUnsafeNew n                  = con `liftM` M.basicUnsafeNew n         \
-; basicInitialize (con v)           = M.basicInitialize v                    \
-; basicUnsafeReplicate n (ty x)     = con `liftM` M.basicUnsafeReplicate n x \
-; basicUnsafeRead (con v) i         = ty `liftM` M.basicUnsafeRead v i       \
-; basicUnsafeWrite (con v) i (ty x) = M.basicUnsafeWrite v i x               \
-; basicClear (con v)                = M.basicClear v                         \
-; basicSet (con v) (ty x)           = M.basicSet v x                         \
-; basicUnsafeCopy (con v1) (con v2) = M.basicUnsafeCopy v1 v2                \
-; basicUnsafeMove (con v1) (con v2) = M.basicUnsafeMove v1 v2                \
-; basicUnsafeGrow (con v) n         = con `liftM` M.basicUnsafeGrow v n      \
-}
-#define newtypeVectorMono(ty,con,mcon) \
-instance G.Vector Vector ty where { \
-; {-# INLINE basicUnsafeFreeze  #-}                                       \
-; {-# INLINE basicUnsafeThaw    #-}                                       \
-; {-# INLINE basicLength        #-}                                       \
-; {-# INLINE basicUnsafeSlice   #-}                                       \
-; {-# INLINE basicUnsafeIndexM  #-}                                       \
-; {-# INLINE elemseq            #-}                                       \
-; basicUnsafeFreeze (mcon v)        = con `liftM` G.basicUnsafeFreeze v   \
-; basicUnsafeThaw (con v)           = mcon `liftM` G.basicUnsafeThaw v    \
-; basicLength (con v)               = G.basicLength v                     \
-; basicUnsafeSlice i n (con v)      = con $ G.basicUnsafeSlice i n v      \
-; basicUnsafeIndexM (con v) i       = ty `liftM` G.basicUnsafeIndexM v i  \
-; basicUnsafeCopy (mcon mv) (con v) = G.basicUnsafeCopy mv v              \
-; elemseq _ (ty a)                  = seq a                               \
-}
-#define deriveNewtypeMonoInstances(ty,rep,con,mcon)\
-newtype instance MVector s ty = mcon (MVector s rep) ;\
-newtype instance Vector    ty = con  (Vector rep)    ;\
-instance Unbox ty                                    ;\
-newtypeMVectorMono(ty, mcon)                         ;\
-newtypeVectorMono(ty, con, mcon)
-
-
-deriveNewtypeMonoInstances(Any, Bool, V_Any, MV_Any)
-deriveNewtypeMonoInstances(All, Bool, V_All, MV_All)
-
+deriveNewtypeInstances((), Any, Bool, Any, V_Any, MV_Any)
+deriveNewtypeInstances((), All, Bool, All, V_All, MV_All)
 
 -- -------
 -- Const
 -- -------
 
-newtype instance MVector s (Const a b) = MV_Const (MVector s a)
-newtype instance Vector (Const a b) = V_Const (Vector a)
-
-instance (Unbox a) => M.MVector MVector (Const a b) where
-  {-# INLINE basicLength #-}
-  {-# INLINE basicUnsafeSlice #-}
-  {-# INLINE basicOverlaps #-}
-  {-# INLINE basicUnsafeNew #-}
-  {-# INLINE basicInitialize #-}
-  {-# INLINE basicUnsafeReplicate #-}
-  {-# INLINE basicUnsafeRead #-}
-  {-# INLINE basicUnsafeWrite #-}
-  {-# INLINE basicClear #-}
-  {-# INLINE basicSet #-}
-  {-# INLINE basicUnsafeCopy #-}
-  {-# INLINE basicUnsafeGrow #-}
-  basicLength (MV_Const v) = M.basicLength v
-  basicUnsafeSlice i n (MV_Const v) = MV_Const $ M.basicUnsafeSlice i n v
-  basicOverlaps (MV_Const v1) (MV_Const v2) = M.basicOverlaps v1 v2
-  basicUnsafeNew n = MV_Const `liftM` M.basicUnsafeNew n
-  basicInitialize (MV_Const v) = M.basicInitialize v
-  basicUnsafeReplicate n (Const x) = MV_Const `liftM` M.basicUnsafeReplicate n x
-  basicUnsafeRead (MV_Const v) i = Const `liftM` M.basicUnsafeRead v i
-  basicUnsafeWrite (MV_Const v) i (Const x) = M.basicUnsafeWrite v i x
-  basicClear (MV_Const v) = M.basicClear v
-  basicSet (MV_Const v) (Const x) = M.basicSet v x
-  basicUnsafeCopy (MV_Const v1) (MV_Const v2) = M.basicUnsafeCopy v1 v2
-  basicUnsafeMove (MV_Const v1) (MV_Const v2) = M.basicUnsafeMove v1 v2
-  basicUnsafeGrow (MV_Const v) n = MV_Const `liftM` M.basicUnsafeGrow v n
-
-instance (Unbox a) => G.Vector Vector (Const a b) where
-  {-# INLINE basicUnsafeFreeze #-}
-  {-# INLINE basicUnsafeThaw #-}
-  {-# INLINE basicLength #-}
-  {-# INLINE basicUnsafeSlice #-}
-  {-# INLINE basicUnsafeIndexM #-}
-  {-# INLINE elemseq #-}
-  basicUnsafeFreeze (MV_Const v) = V_Const `liftM` G.basicUnsafeFreeze v
-  basicUnsafeThaw (V_Const v) = MV_Const `liftM` G.basicUnsafeThaw v
-  basicLength (V_Const v) = G.basicLength v
-  basicUnsafeSlice i n (V_Const v) = V_Const $ G.basicUnsafeSlice i n v
-  basicUnsafeIndexM (V_Const v) i = Const `liftM` G.basicUnsafeIndexM v i
-  basicUnsafeCopy (MV_Const mv) (V_Const v) = G.basicUnsafeCopy mv v
-  elemseq _ (Const a) = G.elemseq (undefined :: Vector a) a
-
-instance (Unbox a) => Unbox (Const a b)
+deriveNewtypeInstances(Unbox a, Const a b, a, Const, V_Const, MV_Const)
 
 -- ---
 -- Alt
 -- ---
 
 #if MIN_VERSION_base(4,8,0)
-newtype instance MVector s (Alt f a) = MV_Alt (MVector s (f a))
-newtype instance Vector (Alt f a) = V_Alt (Vector (f a))
-
-instance (Unbox (f a)) => M.MVector MVector (Alt f a) where
-  {-# INLINE basicLength #-}
-  {-# INLINE basicUnsafeSlice #-}
-  {-# INLINE basicOverlaps #-}
-  {-# INLINE basicUnsafeNew #-}
-  {-# INLINE basicInitialize #-}
-  {-# INLINE basicUnsafeReplicate #-}
-  {-# INLINE basicUnsafeRead #-}
-  {-# INLINE basicUnsafeWrite #-}
-  {-# INLINE basicClear #-}
-  {-# INLINE basicSet #-}
-  {-# INLINE basicUnsafeCopy #-}
-  {-# INLINE basicUnsafeGrow #-}
-  basicLength (MV_Alt v) = M.basicLength v
-  basicUnsafeSlice i n (MV_Alt v) = MV_Alt $ M.basicUnsafeSlice i n v
-  basicOverlaps (MV_Alt v1) (MV_Alt v2) = M.basicOverlaps v1 v2
-  basicUnsafeNew n = MV_Alt `liftM` M.basicUnsafeNew n
-  basicInitialize (MV_Alt v) = M.basicInitialize v
-  basicUnsafeReplicate n (Alt x) = MV_Alt `liftM` M.basicUnsafeReplicate n x
-  basicUnsafeRead (MV_Alt v) i = Alt `liftM` M.basicUnsafeRead v i
-  basicUnsafeWrite (MV_Alt v) i (Alt x) = M.basicUnsafeWrite v i x
-  basicClear (MV_Alt v) = M.basicClear v
-  basicSet (MV_Alt v) (Alt x) = M.basicSet v x
-  basicUnsafeCopy (MV_Alt v1) (MV_Alt v2) = M.basicUnsafeCopy v1 v2
-  basicUnsafeMove (MV_Alt v1) (MV_Alt v2) = M.basicUnsafeMove v1 v2
-  basicUnsafeGrow (MV_Alt v) n = MV_Alt `liftM` M.basicUnsafeGrow v n
-
-instance (Unbox (f a)) => G.Vector Vector (Alt f a) where
-  {-# INLINE basicUnsafeFreeze #-}
-  {-# INLINE basicUnsafeThaw #-}
-  {-# INLINE basicLength #-}
-  {-# INLINE basicUnsafeSlice #-}
-  {-# INLINE basicUnsafeIndexM #-}
-  {-# INLINE elemseq #-}
-  basicUnsafeFreeze (MV_Alt v) = V_Alt `liftM` G.basicUnsafeFreeze v
-  basicUnsafeThaw (V_Alt v) = MV_Alt `liftM` G.basicUnsafeThaw v
-  basicLength (V_Alt v) = G.basicLength v
-  basicUnsafeSlice i n (V_Alt v) = V_Alt $ G.basicUnsafeSlice i n v
-  basicUnsafeIndexM (V_Alt v) i = Alt `liftM` G.basicUnsafeIndexM v i
-  basicUnsafeCopy (MV_Alt mv) (V_Alt v) = G.basicUnsafeCopy mv v
-  elemseq _ (Alt a) = G.elemseq (undefined :: Vector a) a
-
-instance (Unbox (f a)) => Unbox (Alt f a)
+deriveNewtypeInstances(Unbox (f a), Alt f a, f a, Alt, V_Alt, MV_Alt)
 #endif
 
 -- -------
@@ -718,52 +569,7 @@ instance (Unbox (f a)) => Unbox (Alt f a)
 -- -------
 
 #if MIN_VERSION_base(4,9,0)
-newtype instance MVector s (Compose f g a) = MV_Compose (MVector s (f (g a)))
-newtype instance Vector (Compose f g a) = V_Compose (Vector (f (g a)))
-
-instance (Unbox (f (g a))) => M.MVector MVector (Compose f g a) where
-  {-# INLINE basicLength #-}
-  {-# INLINE basicUnsafeSlice #-}
-  {-# INLINE basicOverlaps #-}
-  {-# INLINE basicUnsafeNew #-}
-  {-# INLINE basicInitialize #-}
-  {-# INLINE basicUnsafeReplicate #-}
-  {-# INLINE basicUnsafeRead #-}
-  {-# INLINE basicUnsafeWrite #-}
-  {-# INLINE basicClear #-}
-  {-# INLINE basicSet #-}
-  {-# INLINE basicUnsafeCopy #-}
-  {-# INLINE basicUnsafeGrow #-}
-  basicLength (MV_Compose v) = M.basicLength v
-  basicUnsafeSlice i n (MV_Compose v) = MV_Compose $ M.basicUnsafeSlice i n v
-  basicOverlaps (MV_Compose v1) (MV_Compose v2) = M.basicOverlaps v1 v2
-  basicUnsafeNew n = MV_Compose `liftM` M.basicUnsafeNew n
-  basicInitialize (MV_Compose v) = M.basicInitialize v
-  basicUnsafeReplicate n (Compose x) = MV_Compose `liftM` M.basicUnsafeReplicate n x
-  basicUnsafeRead (MV_Compose v) i = Compose `liftM` M.basicUnsafeRead v i
-  basicUnsafeWrite (MV_Compose v) i (Compose x) = M.basicUnsafeWrite v i x
-  basicClear (MV_Compose v) = M.basicClear v
-  basicSet (MV_Compose v) (Compose x) = M.basicSet v x
-  basicUnsafeCopy (MV_Compose v1) (MV_Compose v2) = M.basicUnsafeCopy v1 v2
-  basicUnsafeMove (MV_Compose v1) (MV_Compose v2) = M.basicUnsafeMove v1 v2
-  basicUnsafeGrow (MV_Compose v) n = MV_Compose `liftM` M.basicUnsafeGrow v n
-
-instance (Unbox (f (g a))) => G.Vector Vector (Compose f g a) where
-  {-# INLINE basicUnsafeFreeze #-}
-  {-# INLINE basicUnsafeThaw #-}
-  {-# INLINE basicLength #-}
-  {-# INLINE basicUnsafeSlice #-}
-  {-# INLINE basicUnsafeIndexM #-}
-  {-# INLINE elemseq #-}
-  basicUnsafeFreeze (MV_Compose v) = V_Compose `liftM` G.basicUnsafeFreeze v
-  basicUnsafeThaw (V_Compose v) = MV_Compose `liftM` G.basicUnsafeThaw v
-  basicLength (V_Compose v) = G.basicLength v
-  basicUnsafeSlice i n (V_Compose v) = V_Compose $ G.basicUnsafeSlice i n v
-  basicUnsafeIndexM (V_Compose v) i = Compose `liftM` G.basicUnsafeIndexM v i
-  basicUnsafeCopy (MV_Compose mv) (V_Compose v) = G.basicUnsafeCopy mv v
-  elemseq _ (Compose a) = G.elemseq (undefined :: Vector a) a
-
-instance (Unbox (f (g a))) => Unbox (Compose f g a)
+deriveNewtypeInstances(Unbox (f (g a)), Compose f g a, f (g a), Compose, V_Compose, MV_Compose)
 #endif
 
 -- ------
```

Shimuuar commented:
@RyanGlScott I've applied your patch.

RyanGlScott commented:
I'd say this is as streamlined as we're going to get, so LGTM.

Thanks, @Shimuuar!
